src/keyboardNavigation.ts

```diff
@@ -0,0 +1,82 @@
+    if (!parent || !parent.parentElement) return initial;
+    const parentAndSiblings = [...parent.parentElement.children];
+    if (isLastChild(parent, parentAndSiblings)) {
+        return findNearestParentWithAdditionalLeaves(parent, initial);
+    } else {
+        return getNextSibling(parent, parentAndSiblings);
+    }
+}
+
+export const navigateUp = (path: string, shadowRoot: ShadowRoot): void => {
+    const currentFocus = <HTMLElement>shadowRoot.querySelector(`[data-path="${path}"]`);
+    if (!currentFocus || !currentFocus.parentNode) return;
+    const currentFocusAndSiblings = [...currentFocus.parentNode.children];
+    const previousSibling = getPreviousSibling(currentFocus, currentFocusAndSiblings);
+    if (previousSibling instanceof ShadowRoot) return;
+    const newFocus =
+        !isDeepestVisible(previousSibling) && !isAncestor(path, previousSibling.getAttribute('data-path') || '')
+            ? findDeepestVisibleChild(previousSibling)
+            : previousSibling;
+    if (newFocus) changeFocusedKey(newFocus, currentFocus);
+};
+
+export const navigateDown = (path: string, shadowRoot: ShadowRoot): void => {
+    const currentFocus = <HTMLElement>shadowRoot.querySelector(`[data-path="${path}"]`);
+    if (!currentFocus || !currentFocus.parentNode) return;
+    const currentFocusAndSiblings = [...currentFocus.parentNode.children];
+    const newFocus = findNextElementToFocus(currentFocus, currentFocusAndSiblings);
+    if (newFocus) changeFocusedKey(newFocus, currentFocus);
+};
+
+const findNextElementToFocus = (currentFocus: HTMLElement, currentFocusAndSiblings: Array<Element>) => {
+    if (!isDeepestVisible(currentFocus)) {
+        if (isEmptyObject(currentFocus)) {
+            return findNearestParentWithAdditionalLeaves(currentFocus);
+        } else {
+            return <HTMLElement>currentFocus.querySelector('[part=property]');
+        }
+    } else if (isLastChild(currentFocus, currentFocusAndSiblings)) {
+        return findNearestParentWithAdditionalLeaves(currentFocus) || currentFocus;
+    } else return getNextSibling(currentFocus, currentFocusAndSiblings);
+};
+
+export const navigateToParent = (path: string, shadowRoot: ShadowRoot): void => {
+    const currentFocus = <HTMLElement>shadowRoot.querySelector(`[data-path="${path}"]`);
+    if (!currentFocus || !currentFocus.parentNode || !currentFocus.parentNode.parentNode) return;
+    changeFocusedKey(<HTMLElement>currentFocus.parentNode.parentNode, currentFocus);
+};
+
+const changeFocusedKey = (newFocus: HTMLElement, oldFocus: HTMLElement): void => {
+    const keyToFocus = <HTMLSpanElement>newFocus.querySelector('[part=key]');
+    const keyToBlur = <HTMLSpanElement>oldFocus.querySelector('[part=key]');
+    keyToFocus.setAttribute('tabindex', '0');
+    keyToBlur.setAttribute('tabindex', '-1');
+    keyToFocus.focus();
+};
+
+const getNextSibling = (currentFocus: HTMLElement, currentFocusAndSiblings: Array<Element>): HTMLElement =>
```

Comment:
it leads to a couple extra generated arrays, but what do you think of having these two helpers just take `currentFocus` and create the array themselves? it'll be less bug-prone in the long run

Reply:
I'd love some help doing this in a non-gross typescript way if its possible. I'm definitely partial to the cleanliness of lambdas, and returning defaults feels worse than creating the array in the caller and just being able to check. I've included the commit so you can see my attempt, but dislike the current state enough that I plan on reverting.

Reply:
wanna pair on it sometime tomorrow? i don't immediately have an idea of what to do but i do feel like we can do it better (and both learn a little typescript along the way)

Reply:
Yep, whenever you have time works for me. I'm still a bit confused about the grandparent vs sibling logic split discussion, so I'd like to cover that as well.